### DISCUSSION TITLE
capture to LVGL sample: ov5460 on a STM32H7a3zi

### DISCUSSION THREAD

**Discussion by josuah** [zephyrproject-rtos/zephyr#80371] [Q&A]
@CharlesDias I am trying to use the ov5460 on a STM32H7a3zi. Im using latest Zephyr master but I get 
 video_ov5640: Unsupported pixel format or resolution.

Is there any PR I need for this to run thats not yet in master? Or did I do something wrong? Im using the lvgl sample but removed all lvgl code as I dont have a display.

_Originally posted by @CodingGhost in https://github.com/zephyrproject-rtos/zephyr/issues/74353#issuecomment-2434885948_

**Comment by josuah** [comment]
> Unsupported pixel format or resolution

The pixel format is selected by the application.
The application of the LVGL sample is supporting multiple hardware.
Each hardware support a different subset of formats.
To allow supporting all hardware, users might need to specify the format they wish to use. The one that matches their hardware.

You can do this with this Kconfig using a "fourCC" string such as one listed from [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/video.h#L761).
This option was introduced [here](https://github.com/zephyrproject-rtos/zephyr/pull/77879):
https://github.com/zephyrproject-rtos/zephyr/pull/77879/files#diff-737bb0ca878e20bfaaf493aae9eb42d3de0e40e22a0fd17ac08d826050c1d41bR11

It is possible that this broke some other boards, for which the resolution was not specified.
If you wish to patch it upstream, you can add a line like this on a new `board/stm32h7...conf` file in the samples directory: https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/video/capture/boards
```
CONFIG_VIDEO_PIXEL_FORMAT="RGB8"
```

Or in command line: `west build ... -- -D CONFIG_VIDEO_PIXEL_FORMAT="RGB8"` (I think).

With whichever format works with your particular display.
This sample does not the pixel format conversion (too slow with software), so you need to make sure the format chosen for the camera matches the one of the display.

Maybe it would be good to document that the user needs to select which pixel format to use between the display and camera...

If that does not work, we can continue the discussion here :)
Then depending on the outcome, we would know how we can patch it upstream.
Thank you for trying!

**Reply by CodingGhost** [reply]
Hi, thanks four your quick reply! 
I am only trying to get some data from the camera, Im not even using a display. I removed all the lvgl related configurations and code.
I originally posted this question on #74353 because Im using ov5460@dcmi and it seems like this combination is not yet fully supported. Is this correct? I think I need some PR that is not merged into main yet.
Or did I misunderstand all this and ov5460@dcmi should already work without problems?

Im also wondering if my wiring is correct, as the adafruit module im using has pins D2...D9 instead of D0..D7 and there is an external clock input.
I connected D2-->D0 .. D9-->D7 and set a jumper on the camera that is supposed to use an internal clock. Also pixelclk, hsync and vsync are connected ofcourse.
Does all this seem correct?

**Reply by josuah** [reply]
I think you are right: DVP is not yet upstreamed into Zephyr for OV5640. Only elements that helps with this here and there, but not the end to end flow.

However, the particular error message you encontered was about a *different* issue I believe. It can be interesting to adjust it and see what the next error you encounter is.

**Reply by josuah** [reply]
> I'm also wondering if my wiring is correct, as the adafruit module im using has pins D2...D9 instead of D0..D7 and there is an external clock input.
> I connected D2-->D0 .. D9-->D7 and set a jumper on the camera that is supposed to use an internal clock. Also pixelclk, hsync and vsync are connected ofcourse.

If you have a logic analyzer or oscilloscope, you would be able to check if the signals are looking correct.
For instance, whether there is a clock signal going on, whether there is any data coming through via I2C, are the data pins emitting anything at all or not.

There is typically a XCLK for providing the camera its internal clock, and a PCLK, the pixel clock, that marks the timing of the D0-D7 pins.

Once the internal clock is provided, I2C communication is expected to work.
Once the camera is started, you would receive the data synchronized with the pixel clock, and the HSYNC pusling on every row, and VSYNC pulsing on every frame.

**Reply by CharlesDias** [reply]
Hi @josuah and @CodingGhost,

The DVP support for OV5640 is not upstreamed yet. Please, check the PRs below:
- https://github.com/zephyrproject-rtos/zephyr/pull/76124
- https://github.com/zephyrproject-rtos/zephyr/pull/76143

**Reply by CodingGhost** [reply]
thanks! I cherry picked the commit for the DVP support and now It works a little bit more :D
I get the "Capture started" log line and the test1 but then nothing after.
`	LOG_INF("- Capture started");

	/* Grab video frames */
	while (1) {
		int err;
	LOG_INF("test1");
		err = video_dequeue(video_dev, VIDEO_EP_OUT, &vbuf, K_FOREVER);
		if (err) {
			LOG_ERR("Unable to dequeue video buf");
			return 0;
		}
		LOG_INF("test2");
		//LOG_DBG("Got frame %u! size: %u; timestamp %u ms", frame++, vbuf->bytesused,vbuf->timestamp);

		err = video_enqueue(video_dev, VIDEO_EP_OUT, vbuf);
		if (err) {
			LOG_ERR("Unable to requeue video buf");
			return 0;
		}
	}`

It seems like the vide_dequeue function is hanging. Any Ideas why that could be? 
If not Im going to have to debug into it.

**Reply by CharlesDias** [reply]
Probably, you are run out of vídeo memory. Please, increase the video memory pool.

**Reply by CharlesDias** [reply]
For example, samples/drivers/video/capture_to_lvgl/boards/mini_stm32h743.conf

**Reply by josuah** [reply]
I think this would need a check in the Zephyry API, as very frequent source of confusion: if a buffer enqueued is larger than the total memory size, return NULL: easier to debug a crash than a hang.

https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_VIDEO_BUFFER_POOL_SZ_MAX
https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_VIDEO_BUFFER_POOL_NUM_MAX

**Reply by CodingGhost** [reply]
CONFIG_VIDEO_BUFFER_POOL_SZ_MAX=80000 is the max I can do, if I do more, region RAM overflows. I still have the same problem after doing so.
a quick debug shows that video_stm32_dcmi.c:341 ( *vbuf = k_fifo_get(&data->fifo_out, timeout); ) seems to cause the hanging.
that seems to make sense if its ram related?
Is there anything I can try to reduce ram usage of the video? whats even stored in the buffer? only one frame or multiple or just one line..? as I need to read images with 1MP, if it is a requirement that a full image fits into ram, this cannot work anyways...
I tried changing pixelformat from RGB565 to RGGB8 but this does not work...

**Reply by CharlesDias** [reply]
@CodingGhost , please set up the minimal resolution to test, like 160x120. Also, you have to respect the minimal ratio between AHB/PIXCLK of 2.5. See https://www.st.com/resource/en/product_training/STM32F7_Peripheral_DCMI.pdf

@josuah , the video_dequeue doesn't return if run out of memory. This was something that we talked some months ago on discord.

**Reply by josuah** [reply]
> CONFIG_VIDEO_BUFFER_POOL_SZ_MAX=80000 is the max I can do

This means it requires some strategy to work around the hardware limitations:

- Try to configure `CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=1` (fewer buffers).
- Try to configure the OV5640 to a smaller size.
- Try to free-up memory from elsewhere to make a full frame fit.
- Try to integrate a partial read API inside the DCMI driver, which was introduced by NXP as part of the SmartDMA driver:  [`vbuf.line_offset`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/video.h#L137-L142)
- Try to see if there are other memory banks that can fit a whole frame.
- Use a pixel format that is using fewer bytes per pixel (like bayer as you tried)
- Implement the JPEG mode of the OV5640 (which requires a bit of work) to fit a whole frame as JPEG compressed image
- Always possible to experiment and try other things!

**Reply by CodingGhost** [reply]
OK, so I tried CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=1 and reduced the video_buffer array size from the sample from two to one. now it actually fits. But the behaviour is still the same, it hangs in video_dequeue.
Besides that problem, if I want to take pictures with 1Mpixel there is no way a full frame will fit into ram, so a partial read API would be needed anyways. Is this partial read feature working with every cam / chip? or only NXP hardware and I need to implement it for stm32 ? Is there any documentation to it?

**Reply by josuah** [reply]
> Is this partial read feature working with every cam / chip?

I think this is a feature of the controller associated with the camera: the camera itself does not do any data movement. The MIPI/DVP/... core of the chip does. You would need to implement it for the STM32 DCMI driver I think.

The only documentation for it is this: 

```c
	/** Line offset within frame this buffer represents, from the
	 * beginning of the frame. This offset is given in pixels,
	 * so `line_offset` * `pitch` provides offset from the start of
	 * the frame in bytes.
	 */
	uint16_t line_offset;
```

In other words, the driver needs to be able to handle a `vbuf->size` that is shorter than the full frame length, and to tell where in the frame it is at, `line_offset` can be used, to keep track of the progress through the frame.

For instance for a 480x320 (HVGA) frame size in BGGR8 format (`fmt->pitch == fmt->width` as there is 1 byte per pixel):

- If transmitting from 0th line to the 99th line (100 lines in total), then `vbuf->size` needs to be at least `fmt->pitch * 100` and `vbuf->line_offset` can be set to 0.
- If transmitting from 100th line to the 199th line (100 lines in total), then `vbuf->size` needs to be at least `fmt->pitch * 100` and `vbuf->line_offset` can be set to 100.
- If transmitting from 200th line to the 299th line (100 lines in total), then `vbuf->size` needs to be at least `fmt->pitch * 100` and `vbuf->line_offset` can be set to 200.
- If transmitting from 300th line to the 319th line (20 last lines of the frame), then `vbuf->size` needs to be at least `fmt->pitch * 20` and `vbuf->line_offset` can be set to 300.

Then the application would know that 100 + 100 + 100 + 20 = 320 lines were transferred, which is `fmt->height`: a complete frame.

**Reply by josuah** [reply]
As an example implementation:
https://github.com/zephyrproject-rtos/zephyr/blob/95cc5f53b82cbc87da1e0f8f402bb1430bda9d66/drivers/video/video_mcux_smartdma.c#L87

**Reply by CharlesDias** [reply]
@CodingGhost, 

It is necessary a lot of memory to use DVP with high resolution. Into the DCMI driver is created a memory pool equal to CONFIG_VIDEO_BUFFER_POOL_SZ_MAX that is used to receive the frame via DMA. https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20video_stm32_buffer_pool&type=code

If you are using, for example 320x240 pixels and VIDEO_PIX_FMT_RGB565, it will be necessary 320 * 240 * 2 = 153600 bytes! Considering set up the CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=1, it'll be necessary at least 153600 bytes for the DCMI driver and 153600 bytes for the video buffer in application level.

See the sample https://docs.zephyrproject.org/latest/samples/drivers/video/capture_to_lvgl/README.html#video-capture-to-lvgl

**Reply by CharlesDias** [reply]
In PR https://github.com/zephyrproject-rtos/zephyr/pull/76143 I delete the sram1 and sram2 nodes to increase the sram0. https://github.com/zephyrproject-rtos/zephyr/blob/32b6c74f3d561205fdf043e1d47de861e2e98b40/samples/drivers/video/capture_to_lvgl/boards/stm32h7b3i_dk.overlay

You can try something similar for contiguous memory block.

**Reply by CharlesDias** [reply]
Hi @CodingGhost , did you check the minimal ratio between AHB/PIXCLK of 2.5? See https://www.st.com/resource/en/product_training/STM32F7_Peripheral_DCMI.pdf

**Reply by CodingGhost** [reply]
@CharlesDias I did not change it, I dont even know how I can change it. Where can I set the Pixelclock? Maybe thats missing? 
Is there anything I need to modify from the bare capture to LVGL sample because it originally doesnt use dcmi? I thought not because over at discord you posted that this sample can be used to test the ov5640 PR

**Reply by CharlesDias** [reply]
@CodingGhost , which board are ypu using? The PIXCLK depende on the oscillator used on ov5640 and some register configuration (I don't know which one). 

You can meassure it using the oscilloscope or logic analyzer.  Also, you can try to increase the AHB  frequency.

**Reply by CodingGhost** [reply]
@CharlesDias Im using the Adafruit Ov5640 breakout on a STM32H7a3Zi-q. That makes sense, I thought the pixclk is generated by the Driver, but the pixelclock is generated by the camera, and the STM32 is just reading it.  But wouldnt the ov5640 driver need to be able to set the Register for the Pixelclock speed? I will try to find out how I can change the AHB clock first though.

**Reply by CharlesDias** [reply]
@CodingGhost , check on the schematic of the board if it has a oscillator to ov5640 or if the clock came from the MCO microntroller pin.

To change the AHB frequency, you have to change the device tree configuration, nodes pll and rcc. You can use the STM32CubeMX to check the correct values or the research on datasheet. 😀

**Reply by CodingGhost** [reply]
@CharlesDias I have set a jumper on the camera board that is supposed to generate a clock for the cam. I guess thats what you mean? It also has the pxlclk pin, but if I understand correctly this pin is an output and not an input.
I will check what the clocks currently are to make sure that this is not the problem.

**Comment by CodingGhost** [comment]
@CharlesDias thanks for that Tip! I created one big sram partition and now I can use
CONFIG_VIDEO_BUFFER_POOL_NUM_MAX=2
CONFIG_VIDEO_BUFFER_POOL_SZ_MAX=262144

But the application still hangs, even though im only using 160x120.

**Reply by CodingGhost** [reply]
@CharlesDias I checked with cubeMX and it seems the AHB clock is already maxed out by default.

I have pasted my full demo code here:
https://pastebin.com/dL0fS7EB
and my full dtc here:
https://pastebin.com/wfGJWt37

Can you spot anything that could cause the hanging here? Im out of ideas, But Im not familiar with dcmi anyways...

**Reply by CharlesDias** [reply]
Hi @CodingGhost. Sorry for the delay in my answers. These days are tough. Next week, I hope to be better and have time to dig into this problem.

Have you seen this information? According to the pinouts' documentation, "[...] the XCLK jumper is set to "EXT" (the default), this pin must be driven with a 24MHz square wave from the microcontroller [...]". In this case, to keep this configuration, the microcontroller has to provide the XCLK.You can use the MCO pin of microcontroller as I did here https://github.com/zephyrproject-rtos/zephyr/blob/992d208bfc62a8dfb81d38249e3a3e82813b3c5e/boards/shields/weact_ov5640_cam_module/boards/mini_stm32h743.overlay. 

https://learn.adafruit.com/adafruit-ov5640-camera-breakout/pinouts
![image](https://github.com/user-attachments/assets/efb7d89b-e967-42e0-87a8-361a7214e0f8)

**Reply by CharlesDias** [reply]
@CodingGhost 

Also, pay attention to your wire cable connection and quality. Some months ago, I spent days looking for a problem with the ov5640 sample program, but the issue was two malfunctioning cables.

**Reply by CodingGhost** [reply]
@CharlesDias I have set the jumper to use the internal clock. I can check wiring again, but I think my wiring was ok. And yeah no worries, take your time. Thanks for the help!

**Reply by CharlesDias** [reply]
@CodingGhost do you have debugger and logic analyzer or oscilloscope?

**Reply by CodingGhost** [reply]
I have a debugger and a dso, I think I can use my dso with pulseview.

**Reply by CodingGhost** [reply]
@CharlesDias I checked the wiring again, and this time I found a jumper calbe with a loose connection. replaced it and it instantly works. damn it could have been so easy :D
thanks for all the help anyways! now I think I have to implement JPEG mode, do you have any hint with where to start implementing this?

**Reply by josuah** [reply]
> where to start implementing this?

For the OV5640, the [datasheet](https://cdn.sparkfun.com/datasheets/Sensors/LightImaging/OV5640_datasheet.pdf) is exceptionally available, and even [that appnote distributed to OmniVision FAEs only](https://www.arducam.com/downloads/modules/OV5640/OV5640_Software_app_note_parallel.pdf)!

It contains sequences of registers to convert from one mode to another.

**Reply by CharlesDias** [reply]
> damn it could have been so easy :D

If it were easy, it wouldn't be fun and rewarding! Congrats, @CodingGhost ! 

About JPEG, I didn't have the opportunity to dig into. By the way, the DCMI driver does not support JPEG so far. ASAP I'll start to work on this improvement.
